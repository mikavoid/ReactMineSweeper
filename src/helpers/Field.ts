export type Cell = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type Field = Cell[][];
export type Coord = [number, number];

export const CellState: Record<string, Cell> = {
  empty: 1,
  bomb: 9,
  hidden: 10,
  mark: 11,
  weakMark: 12,
};
