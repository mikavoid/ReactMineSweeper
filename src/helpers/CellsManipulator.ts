import { Cell, Coords, Field } from "./Field";

export const incrementNeighboors = (coords: Coords, field: Field): Field => {
  const items = getNeighboorsItems(coords);

  console.log(Object.values(items));

  for (const item of Object.values(items)) {
    if (checkItemInField(item, field)) {
      const [y, x] = item;
      const cell = field[y][x];

      if (cell < 8) {
        field[y][x] = (cell + 1) as Cell;
      }
    }
  }
  return field;
};

export const getNeighboorsItems = (coords: Coords): Record<string, Coords> => {
  const [y, x] = coords;

  return {
    top: [y - 1, x],
    topRight: [y - 1, x + 1],
    right: [y, x + 1],
    rightBottom: [y + 1, x + 1],
    bottom: [y + 1, x],
    bottomLeft: [y + 1, x - 1],
    left: [y, x - 1],
    leftTop: [y - 1, x - 1],
  };
};

export const checkItemInField = ([y, x]: Coords, field: Field) => {
  return field?.[y]?.[x] !== undefined;
};
