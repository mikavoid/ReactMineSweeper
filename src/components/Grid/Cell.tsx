import React, { FC } from "react";
import styled from "@emotion/styled";

import { Cell as CellType, CellState, Coords } from "@/helpers/Field";
import { useMouseDown } from "@/hooks/useMouseDown";

interface CloseFrameProps {
  mouseDown?: boolean;
}

export const ClosedFrame = styled.div<CloseFrameProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: ${({ mouseDown = false }) =>
    mouseDown ? "transparent" : " white #9e9e9e #9e9e9e white"};
  &:hover {
    filter: brightness(1.1);
  }
`;

const RevealedFrame = styled(ClosedFrame)`
  border-color: #dddddd;
  cursor: default;
  font-size: 1em;
  color: ${({ children }) => colors[children as CellType] ?? transparent};
  &:hover {
    filter: brightness(1);
  }
`;

const Bomb = styled.div`
  border-radius: 50%;
  width: 1vh;
  height: 1vh;
  background-color: #333;
`;

const BombFrame = styled(RevealedFrame)`
  background-color: #ec433c;
`;

const Flag = styled.div`
  width: 0px;
  height: 0px;
  border-top: 0.5vh solid transparent;
  border-bottom: 0.5vh solid transparent;
  border-left: 0.5vh solid #ec433c;
`;

const TransparentFlag = styled(Flag)`
  border-left: 0.5vh solid #f19996;
`;

const transparent = "rgba(0,0,0,0)";
const colors: { [k in CellType]: string } = {
  0: "#000",
  1: "#2a48ec",
  2: "#2bb13d",
  3: "#ec6561",
  4: "#233db7",
  5: "#a6070f",
  6: "#e400af",
  7: "#906a02",
  8: "#fa0707",
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};

export interface CellProps {
  /**
   * Cell status based on thr CellType
   */
  children: CellType;

  /**
   * Cell coordonates
   */
  coords: Coords;

  /**
   * onClick by cell handler
   */
  onClick: (coords: Coords) => void;

  /**
   * onContextMenu by cell handler
   */
  onContextMenu: (coords: Coords) => void;
}

interface ComponentsMapProps {
  children: CellType;
  onClick: (elem: React.MouseEvent<HTMLElement>) => void;
  onContextMenu: (elem: React.MouseEvent<HTMLElement>) => void;
  mouseDown: boolean;
  onMouseDown: () => void;
  onMouseUp: () => void;
  "data-testid"?: string;
  role?: string;
}

const ComponentsMap: FC<ComponentsMapProps> = ({ children, ...rest }) => {
  const nonActiveCellProps = {
    onContextMenu: rest.onContextMenu,
    "data-testid": rest["data-testid"],
    role: rest?.role,
  };

  switch (children) {
    case CellState.empty:
      return <RevealedFrame {...nonActiveCellProps} />;
    case CellState.bomb:
      return (
        <BombFrame {...nonActiveCellProps}>
          <Bomb />
        </BombFrame>
      );
    case CellState.mark:
      return (
        <ClosedFrame {...rest}>
          <Flag />
        </ClosedFrame>
      );
    case CellState.weakMark:
      return (
        <ClosedFrame {...rest}>
          <TransparentFlag />
        </ClosedFrame>
      );
    case CellState.hidden:
      return <ClosedFrame {...rest} />;
    default:
      return <RevealedFrame {...nonActiveCellProps}>{children}</RevealedFrame>;
  }
};

export const checkCellIsActive = (cell: CellType): boolean =>
  [CellState.hidden, CellState.mark, CellState.weakMark].includes(cell);

export const Cell: FC<CellProps> = ({ children, coords, ...rest }) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();
  const isActiveCell = checkCellIsActive(children);

  const onClick = () => rest?.onClick(coords);
  const onContextMenu = (elem: React.MouseEvent<HTMLElement>) => {
    /**
     * Prevent context menu by default
     */
    elem.preventDefault();

    if (isActiveCell) {
      rest.onContextMenu(coords);
    }
  };

  const props = {
    onClick,
    onContextMenu,
    mouseDown,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    "data-testid": `${coords}`,
    role: "cell",
  };

  return <ComponentsMap {...props}>{children}</ComponentsMap>;
};
