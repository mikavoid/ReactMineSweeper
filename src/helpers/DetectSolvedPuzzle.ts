import { Field, CellState, Cell } from "@/helpers/Field";

const { empty: e, bomb: b, hidden: h, mark: m } = CellState;

export type FlagCounter = number;
export const detectSolvedPuzzle = (
  playerField: Field,
  gameField: Field
): [boolean, number] => {
  let flagCounter: FlagCounter = 0;
  const flattenPlayerField = playerField.flat();
  const flattenGameField = gameField.flat();

  const nbHidden = flattenPlayerField.filter(
    (cell: Cell) => cell === h
  )?.length;
  const nbBombs = flattenGameField.filter((cell: Cell) => cell === b)?.length;

  for (let i = 0; i < flattenGameField.length; i++) {
    if (flattenPlayerField[i] === m && flattenGameField[i] === b) {
      flagCounter++;
    }
  }
  return [flagCounter === nbBombs && nbHidden === 0, flagCounter];
};
