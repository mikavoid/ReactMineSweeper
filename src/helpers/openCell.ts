import { getNeighboorsItems, checkItemInField } from "./CellsManipulator";

import { Coords, CellState, Field } from "@/helpers/Field";

export const openCell = (
  [y, x]: Coords,
  playerField: Field,
  gameField: Field
) => {
  const gameCell = gameField[y][x];

  if (gameCell === CellState.bomb) {
    throw new Error("Game Over");
  }

  if (gameCell === CellState.empty) {
    playerField[y][x] = gameCell;

    const items = getNeighboorsItems([y, x]);

    for (const coords of Object.values(items)) {
      if (checkItemInField(coords, gameField)) {
        const [y, x] = coords;
        const gameCell = gameField[y][x];
        const playerCell = playerField[y][x];

        if (gameCell === CellState.empty && playerCell === CellState.hidden) {
          // result[y][x] = gameField[y][x];
          // emptyCells.push(coords);
          playerField = openCell(coords, playerField, gameField);
        }

        if (gameCell < CellState.bomb) {
          playerField[y][x] = gameCell;
        }
      }
    }
  }

  playerField[y][x] = gameCell;

  return playerField;
};
