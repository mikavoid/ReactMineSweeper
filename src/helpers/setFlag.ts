import { getNeighboorsItems, checkItemInField } from "./CellsManipulator";

import { Coords, CellState, Field } from "@/helpers/Field";
import { detectSolvedPuzzle } from "./DetectSolvedPuzzle";

export const setFlag = (
  [y, x]: Coords,
  playerField: Field,
  gameField: Field,
  prevFlagCounter: number,
  bombs: number
): [Field, boolean, number] => {
  const cell = playerField[y][x];
  const { hidden, mark, weakMark } = CellState;

  switch (cell) {
    case mark:
      playerField[y][x] = weakMark;
      break;
    case weakMark:
      playerField[y][x] = hidden;
      break;
    case hidden:
      if (prevFlagCounter < bombs) {
        playerField[y][x] = mark;
      }
      break;
  }

  const [isSolved, flagCounter] = detectSolvedPuzzle(playerField, gameField);

  return [playerField, isSolved, flagCounter];
};
