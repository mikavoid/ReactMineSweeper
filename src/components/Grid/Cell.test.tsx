import React from "react";
import { render, screen, fireEvent, createEvent } from "@testing-library/react";

import { CellState, Coords } from "@/helpers/Field";

import { Cell, checkCellIsActive, ClosedFrame } from "./Cell";

describe("Cell component check", () => {
  const coords: Coords = [1, 1];

  for (let cell = CellState.empty; cell <= CellState.weakMark; cell++) {
    it("Cell renders correctly", () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      const { asFragment } = render(<Cell {...props}>{cell}</Cell>);
      expect(asFragment()).toMatchSnapshot();
    });

    it("Closed Frame renders correctly", () => {
      const { asFragment } = render(<ClosedFrame mouseDown={true} />);
      expect(asFragment()).toMatchSnapshot();
    });
    it(`Check preventDefault contextMenu for cell of type ${cell}`, () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);

      const cellComp = screen.getByTestId(`${coords}`);

      const contextMenuEvent = createEvent.contextMenu(cellComp);
      fireEvent(cellComp, contextMenuEvent);

      expect(contextMenuEvent.defaultPrevented).toBe(true);
    });

    it("onClick and onContextMenu handler should be called for active cells", () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);

      const cellComp = screen.getByTestId(`${coords}`);

      fireEvent.click(cellComp);
      if (checkCellIsActive(cell)) {
        expect(props.onClick).toHaveBeenCalled();
      } else {
        expect(props.onClick).not.toHaveBeenCalled();
      }

      fireEvent.contextMenu(cellComp);
      if (checkCellIsActive(cell)) {
        expect(props.onContextMenu).toHaveBeenCalled();
      } else {
        expect(props.onContextMenu).not.toHaveBeenCalled();
      }
    });
  }
});
