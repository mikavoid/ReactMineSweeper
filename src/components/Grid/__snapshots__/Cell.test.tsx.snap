// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell component check Cell renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1917pih"
    data-testid="0_1,1"
  />
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 2`] = `
<DocumentFragment>
  <div
    class="css-uxqpjb"
    data-testid="1_1,1"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 3`] = `
<DocumentFragment>
  <div
    class="css-le8ypb"
    data-testid="2_1,1"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 4`] = `
<DocumentFragment>
  <div
    class="css-6w8km6"
    data-testid="3_1,1"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 5`] = `
<DocumentFragment>
  <div
    class="css-urf6n9"
    data-testid="4_1,1"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 6`] = `
<DocumentFragment>
  <div
    class="css-gbjhmc"
    data-testid="5_1,1"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 7`] = `
<DocumentFragment>
  <div
    class="css-1g09j0b"
    data-testid="6_1,1"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 8`] = `
<DocumentFragment>
  <div
    class="css-zxs4qy"
    data-testid="7_1,1"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 9`] = `
<DocumentFragment>
  <div
    class="css-1vqe7pk"
    data-testid="8_1,1"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 10`] = `
<DocumentFragment>
  <div
    class="css-a1md7a"
    data-testid="9_1,1"
  >
    <div
      class="css-3tubcw"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 11`] = `
<DocumentFragment>
  <div
    class="css-so49sq"
    data-testid="10_1,1"
  />
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 12`] = `
<DocumentFragment>
  <div
    class="css-so49sq"
    data-testid="11_1,1"
  >
    <div
      class="css-1w1u3hh"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 13`] = `
<DocumentFragment>
  <div
    class="css-so49sq"
    data-testid="12_1,1"
  >
    <div
      class="css-1wbzzpj"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 2`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 3`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 4`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 5`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 6`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 7`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 8`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 9`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 10`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 11`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 12`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 13`] = `
<DocumentFragment>
  <div
    class="css-j032ne"
  />
</DocumentFragment>
`;
